src/widget/wsingletoncontainer.cpp

```diff
@@ -0,0 +1,55 @@
+#include "widget/wsingletoncontainer.h"
+
+#include <QLayout>
+
+#include "util/assert.h"
+#include "widget/wlibrary.h"
+
+// static
+WSingletonContainer::WidgetMap WSingletonContainer::m_singletons;
+
+// static
+WSingletonContainer* WSingletonContainer::getSingleton(
+        QString objectname, QWidget* pParent) {
+   WidgetMap::const_iterator widget_it = m_singletons.find(objectname);
+    if (widget_it == m_singletons.end()) {
+        qWarning() << "ERROR: Asked for an unknown singleton widget:"
+                   << objectname;
+        return NULL;
+    }
+    return new WSingletonContainer(objectname, pParent);
+}
+
+WSingletonContainer::WSingletonContainer(QString objectname, QWidget* pParent)
+        : WWidgetGroup(pParent), m_pWidget(NULL) {
+    WidgetMap::const_iterator widget_it = m_singletons.find(objectname);
+    DEBUG_ASSERT(widget_it != m_singletons.end());
```

Comment:
That case is handled in the factory function.  Since the constructor is private, this should never happen so this is a good place for a DEBUG_ASSERT (it would indicate some sort of crazy threading problem, so a crash is a good option)